compiler/qsc/src/incremental.rs

```diff
@@ -0,0 +1,199 @@
+
+        let pass_errors = self.passes.run_default_passes(
+            &mut increment.hir,
+            &mut unit.assigner,
+            core,
+            PackageType::Lib,
+        );
+
+        if !pass_errors.is_empty() {
+            return Err(into_errors_with_source(pass_errors, &unit.sources));
+        }
+
+        Ok(increment)
+    }
+
+    /// Compiles an entry expression.
+    ///
+    /// This method returns the AST and HIR packages that were created as a result of
+    /// the compilation, however it does *not* update the current compilation.
+    ///
+    /// The caller can use the returned packages to perform passes,
+    /// get information about the newly added items, or do other modifications.
+    /// It is then the caller's responsibility to merge
+    /// these packages into the current `CompileUnit` using the `update()` method.
+    pub fn compile_expr(&mut self, expr: &str) -> Result<Increment, Errors> {
+        let (core, unit) = self.store.get_open_mut();
+
+        let mut increment = self
+            .frontend
+            .compile_expr(unit, "<entry>", expr)
+            .map_err(into_errors)?;
+
+        let pass_errors = self.passes.run_default_passes(
+            &mut increment.hir,
+            &mut unit.assigner,
+            core,
+            PackageType::Lib,
+        );
+
+        if !pass_errors.is_empty() {
+            return Err(into_errors_with_source(pass_errors, &unit.sources));
+        }
+
+        Ok(increment)
+    }
+
+    /// Updates the current compilation with the AST and HIR packages,
+    /// and any associated context, returned from a previous incremental compilation.
+    pub fn update(&mut self, new: Increment) {
+        let (_, unit) = self.store.get_open_mut();
+
+        // Extend the AST and HIR packages
+        unit.ast.package.extend(new.ast.package);
+        unit.package.extend(new.hir);
+
+        // The new `Increment` will contain the names and tys
+        // from the original package as well, so just
+        // replace the current tables instead of extending.
+        unit.ast.names = new.ast.names;
+        unit.ast.tys = new.ast.tys;
+    }
+
+    /// Returns a reference to the underlying package store.
+    #[must_use]
+    pub fn package_store(&self) -> &PackageStore {
+        self.store.package_store()
+    }
+
+    /// Returns ID of the current `CompileUnit`.
+    #[must_use]
+    pub fn package_id(&self) -> PackageId {
+        self.store.open_package_id()
+    }
+
+    /// Returns the ID of the source package created from the sources
+    /// passed in during inital creation.
+    #[must_use]
+    pub fn source_package_id(&self) -> PackageId {
+        self.source_package_id
+    }
+
+    /// Consumes the incremental compiler and returns an immutable package store.
+    /// This method can be used to finalize the compilation.
+    #[must_use]
+    pub fn into_package_store(self) -> (PackageStore, PackageId) {
+        self.store.into_package_store()
+    }
```

Comment:
nitpick: I like this from an API completeness perspective, but I wonder when we'd actually use it (right now it's not called, so it's nearly dead code but is not flagged as such because it's marked as `pub` and could technically be called by something outside the crate). I think I'm fine keeping it, but some part of me still flags this as unused code that can be safely deleted. All our incremental scenarios are interactive (Python, Jupyter, REPL) and don't really need a concept of finalizing the compilation, even if it makes sense on paper. Perhaps there is an anticipated scenario I'm not seeing?

Reply:
Ah! It's used for the language service. The language service requires a "finished" compilation. I could certainly leave it out until _that_ PR, but I figured showing the "complete" api up front would be helpful .

Reply:
That sounds good enough to me! Glad to know a usage is coming down the pipe.